### DISCUSSION TITLE
When following documentation guide: TweepError: Failed to send request: Only unicode objects are escapable. Got <generator object to_unicode.<locals>.<genexpr> at 0x000001E0C87505F0> of type <class 'generator'>.

### DISCUSSION THREAD

**Discussion by Nowado** [tweepy/tweepy#1506] [Questions] [answered]
I'm trying to follow basic guide from documentation at https://tweepy.readthedocs.io/en/latest/getting_started.html
Code fails at second step. I'm running code in jupyter, in case that's relevant.

```
auth = tweepy.OAuthHandler(API_KEY, API_SECRET)
auth.set_access_token(ACCESS_TOKEN, ACCESS_TOKEN_SECRET)

api = tweepy.API(auth)
```

I provide API key, API secret key, Access token, Access token secret as consumer_key, consumer_secret, access_token, access_token_secret respectively. This (seemingly) works.

`user = api.get_user('twitter')
`fails returning error from the title. 

> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> E:\anaconda3\envs\Twitty\lib\site-packages\tweepy\binder.py in execute(self)
>     183                 try:
> --> 184                     resp = self.session.request(self.method,
>     185                                                 full_url,
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\requests\sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
>     515         )
> --> 516         prep = self.prepare_request(req)
>     517 
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\requests\sessions.py in prepare_request(self, request)
>     448         p = PreparedRequest()
> --> 449         p.prepare(
>     450             method=request.method.upper(),
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\requests\models.py in prepare(self, method, url, headers, files, data, params, auth, cookies, hooks, json)
>     317         self.prepare_body(data, files, json)
> --> 318         self.prepare_auth(auth, url)
>     319 
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\requests\models.py in prepare_auth(self, auth, url)
>     548             # Allow auth to make its changes.
> --> 549             r = auth(self)
>     550 
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\requests_oauthlib\oauth1_auth.py in __call__(self, r)
>     107             # Omit body data in the signing of non form-encoded requests
> --> 108             r.url, headers, _ = self.client.sign(
>     109                 unicode(r.url), unicode(r.method), None, r.headers
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\__init__.py in sign(self, uri, http_method, body, headers, realm)
>     312         request.oauth_params.append(
> --> 313             ('oauth_signature', self.get_oauth_signature(request)))
>     314 
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\__init__.py in get_oauth_signature(self, request)
>     126 
> --> 127         uri, headers, body = self._render(request)
>     128 
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\__init__.py in _render(self, request, formencode, realm)
>     208         if self.signature_type == SIGNATURE_TYPE_AUTH_HEADER:
> --> 209             headers = parameters.prepare_headers(
>     210                 request.oauth_params, request.headers, realm=realm)
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\utils.py in wrapper(params, *args, **kwargs)
>      31         params = filter_oauth_params(params)
> ---> 32         return target(params, *args, **kwargs)
>      33 
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\parameters.py in prepare_headers(oauth_params, headers, realm)
>      58         escaped_name = utils.escape(oauth_parameter_name)
> ---> 59         escaped_value = utils.escape(value)
>      60 
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\utils.py in escape(u)
>      55     if not isinstance(u, unicode_type):
> ---> 56         raise ValueError('Only unicode objects are escapable. ' +
>      57                          'Got %r of type %s.' % (u, type(u)))
> 
> ValueError: Only unicode objects are escapable. Got <generator object to_unicode.<locals>.<genexpr> at 0x000001E0C87505F0> of type <class 'generator'>.
> 
> During handling of the above exception, another exception occurred:
> 
> TweepError                                Traceback (most recent call last)
> <ipython-input-43-362d53fd4374> in <module>
> ----> 1 user = api.get_user('twitter')
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\tweepy\binder.py in _call(*args, **kwargs)
>     251                 return method
>     252             else:
> --> 253                 return method.execute()
>     254         finally:
>     255             method.session.close()
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\tweepy\binder.py in execute(self)
>     190                                                 proxies=self.api.proxy)
>     191                 except Exception as e:
> --> 192                     six.reraise(TweepError, TweepError('Failed to send request: %s' % e), sys.exc_info()[2])
>     193 
>     194                 rem_calls = resp.headers.get('x-rate-limit-remaining')
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\six.py in reraise(tp, value, tb)
>     700                 value = tp()
>     701             if value.__traceback__ is not tb:
> --> 702                 raise value.with_traceback(tb)
>     703             raise value
>     704         finally:
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\tweepy\binder.py in execute(self)
>     182                 # Execute request
>     183                 try:
> --> 184                     resp = self.session.request(self.method,
>     185                                                 full_url,
>     186                                                 data=self.post_data,
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\requests\sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
>     514             hooks=hooks,
>     515         )
> --> 516         prep = self.prepare_request(req)
>     517 
>     518         proxies = proxies or {}
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\requests\sessions.py in prepare_request(self, request)
>     447 
>     448         p = PreparedRequest()
> --> 449         p.prepare(
>     450             method=request.method.upper(),
>     451             url=request.url,
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\requests\models.py in prepare(self, method, url, headers, files, data, params, auth, cookies, hooks, json)
>     316         self.prepare_cookies(cookies)
>     317         self.prepare_body(data, files, json)
> --> 318         self.prepare_auth(auth, url)
>     319 
>     320         # Note that prepare_auth must be last to enable authentication schemes
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\requests\models.py in prepare_auth(self, auth, url)
>     547 
>     548             # Allow auth to make its changes.
> --> 549             r = auth(self)
>     550 
>     551             # Update self to reflect the auth changes.
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\requests_oauthlib\oauth1_auth.py in __call__(self, r)
>     106         else:
>     107             # Omit body data in the signing of non form-encoded requests
> --> 108             r.url, headers, _ = self.client.sign(
>     109                 unicode(r.url), unicode(r.method), None, r.headers
>     110             )
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\__init__.py in sign(self, uri, http_method, body, headers, realm)
>     311         # generate the signature
>     312         request.oauth_params.append(
> --> 313             ('oauth_signature', self.get_oauth_signature(request)))
>     314 
>     315         # render the signed request and return it
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\__init__.py in get_oauth_signature(self, request)
>     125                                             self.resource_owner_secret)
>     126 
> --> 127         uri, headers, body = self._render(request)
>     128 
>     129         collected_params = signature.collect_parameters(
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\__init__.py in _render(self, request, formencode, realm)
>     207         # I'm not sure how I feel about it.
>     208         if self.signature_type == SIGNATURE_TYPE_AUTH_HEADER:
> --> 209             headers = parameters.prepare_headers(
>     210                 request.oauth_params, request.headers, realm=realm)
>     211         elif self.signature_type == SIGNATURE_TYPE_BODY and request.decoded_body is not None:
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\utils.py in wrapper(params, *args, **kwargs)
>      30     def wrapper(params, *args, **kwargs):
>      31         params = filter_oauth_params(params)
> ---> 32         return target(params, *args, **kwargs)
>      33 
>      34     wrapper.__doc__ = target.__doc__
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\parameters.py in prepare_headers(oauth_params, headers, realm)
>      57         # .. _`Section 3.6`: https://tools.ietf.org/html/rfc5849#section-3.6
>      58         escaped_name = utils.escape(oauth_parameter_name)
> ---> 59         escaped_value = utils.escape(value)
>      60 
>      61         # 2.  Each parameter's name is immediately followed by an "=" character
> 
> E:\anaconda3\envs\Twitty\lib\site-packages\oauthlib\oauth1\rfc5849\utils.py in escape(u)
>      54     """
>      55     if not isinstance(u, unicode_type):
> ---> 56         raise ValueError('Only unicode objects are escapable. ' +
>      57                          'Got %r of type %s.' % (u, type(u)))
>      58     # Letters, digits, and the characters '_.-' are already treated as safe
> 
> TweepError: Failed to send request: Only unicode objects are escapable. Got <generator object to_unicode.<locals>.<genexpr> at 0x000001E0C87505F0> of type <class 'generator'>.

Full log above.

I tried googling around and found multiple answers, including older issues in this repo, but none of them work or address the issue (somehow). I'd like to imagine following first steps of example from documentation I shouldn't be able to screw up too much.

What can be wrong, how to fix it?

**Comment by Harmon758** [comment, marked as answer]
You're passing a generator as one or more of the credentials. They need to be strings or bytes.

For code block usage, see https://docs.github.com/en/free-pro-team@latest/github/writing-on-github/creating-and-highlighting-code-blocks.

**Comment by harr1424** [comment]
I am receiving this error when using a Docker container, @Nowado did you determine how you were passing a generator type as parameter? I've confirmed mine should all be interpreted as strings by printing their type. The code below runs fine locally but throes the error in this post's title when executed in a container: 

```
auth = tweepy.OAuth1UserHandler(API_KEY, API_KEY_SECRET, ACCESS_TOKEN, ACCESS_TOKEN_SECRET)
```

**Reply by Nowado** [reply]
I don't remember, sorry : (
